# Sistem Rekomendasi Game mengunakan metode Content Base Filtering

# Pertanyaan Bisnis
1. Bagaimana merekomendasikan game kepada pengguna berdasarkan kesukaan genre yang sama
2. Bagaimana merekomendasikan publisher game yang sama kepada pengguna

#Link dataset
kaagle : https://www.kaggle.com/datasets/uuratl/metacritic-games-12-23-2024

# Persiapan Library

In [3]:
import pandas as pd
import numpy as np
import kagglehub
import re
from kagglehub import KaggleDatasetAdapter
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

# Data Understanding

In [4]:
# Set the path to the file you'd like to load
file_path = "12_23_2024_detailed_data.csv"

df = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "uuratl/metacritic-games-12-23-2024",
  file_path,
)
df.head()

<ipython-input-4-b317389649f5>:4: DeprecationWarning: load_dataset is deprecated and will be removed in a future version.
  df = kagglehub.load_dataset(


100%|██████████| 3.04M/3.04M [00:00<00:00, 108MB/s]


Unnamed: 0                                  name  \
0           0  The Legend of Zelda: Ocarina of Time   
1           1                           SoulCalibur   
2           2                   Grand Theft Auto IV   
3           3                    Super Mario Galaxy   
4           4                  Super Mario Galaxy 2   

      metacritic_review_count metacritic_review_score  \
0  Based on 22 Critic Reviews                      99   
1  Based on 24 Critic Reviews                      98   
2  Based on 86 Critic Reviews                      98   
3  Based on 73 Critic Reviews                      97   
4  Based on 87 Critic Reviews                      97   

              user_review_count user_review_score  \
0  Based on 10,273 User Ratings               9.1   
1     Based on 574 User Ratings               7.8   
2   Based on 4,462 User Ratings               8.3   
3   Based on 4,027 User Ratings               9.1   
4   Based on 2,857 User Ratings               9.0   

                                 developer                  publisher  \
0            Developer:           Nintendo        Publisher: Nintendo   
1               Developer:           Namco           Publisher: Namco   
2      Developer:           Rockstar North  Publisher: Rockstar Games   
3            Developer:           Nintendo        Publisher: Nintendo   
4  Developer:           Nintendo EAD Tokyo        Publisher: Nintendo   

                                        platforms                 genres  \
0                                 ['Nintendo 64']  ['Open-World Action']   
1  ['Dreamcast', 'iOS (iPhone/iPad)', 'Xbox 360']        ['3D Fighting']   
2             ['PlayStation 3', 'Xbox 360', 'PC']  ['Open-World Action']   
3                                         ['Wii']      ['3D Platformer']   
4                                         ['Wii']      ['3D Platformer']   

                                             esrb  must_play  
0  Rated E                           for Everyone          1  
1      Rated T                           for Teen          1  
2    Rated M                           for Mature          1  
3  Rated E                           for Everyone          1  
4  Rated E                           for Everyone          1

Mengambil dataset menggunakan library bawaan kaagle, agar projek bisa di running tanpa perlu input dataset


---
KaggleDatasetAdapter.PANDAS akan memanggil fungsi pandas
file_path, mengambil nama file dari link dataset


# Univariate Exploratory Data Analysis

## melihat dimensi data

In [5]:
df.shape

(13390, 12)

Melihat baris dan kolom yang dimiliki dataset

## Melihat informasi dataset

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13390 entries, 0 to 13389
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Unnamed: 0               13390 non-null  int64 
 1   name                     13390 non-null  object
 2   metacritic_review_count  13384 non-null  object
 3   metacritic_review_score  13390 non-null  object
 4   user_review_count        11833 non-null  object
 5   user_review_score        13390 non-null  object
 6   developer                13380 non-null  object
 7   publisher                13380 non-null  object
 8   platforms                13390 non-null  object
 9   genres                   13390 non-null  object
 10  esrb                     11287 non-null  object
 11  must_play                13390 non-null  int64 
dtypes: int64(2), object(10)
memory usage: 1.2+ MB


Dataset mentah memiliki total 13390 baris, kebanyakan bertype object, dan dua kolom bertype int64

# Data Preparation

## Mengatasi missing value

In [7]:
df.isnull().sum()

Unnamed: 0                    0
name                          0
metacritic_review_count       6
metacritic_review_score       0
user_review_count          1557
user_review_score             0
developer                    10
publisher                    10
platforms                     0
genres                        0
esrb                       2103
must_play                     0
dtype: int64

Terdapat 2 kolom yang memiliki missing value, yaitu user_review_count dan esrb.


---
mengunakan perintah isnull untuk mengetahui kolom yang memiliki nilai null, dan sum() untuk menjumlahkan nilai null dalam satu kolom


In [8]:
df_clean = df.dropna()
df_clean

Unnamed: 0                                     name  \
0               0     The Legend of Zelda: Ocarina of Time   
1               1                              SoulCalibur   
2               2                      Grand Theft Auto IV   
3               3                       Super Mario Galaxy   
4               4                     Super Mario Galaxy 2   
...           ...                                      ...   
13384       13384                Ride to Hell: Retribution   
13385       13385                             SPOGS Racing   
13386       13386  Double Dragon II: Wander of the Dragons   
13387       13387                   Vroom in the Night Sky   
13388       13388                                    Yaris   

          metacritic_review_count metacritic_review_score  \
0      Based on 22 Critic Reviews                      99   
1      Based on 24 Critic Reviews                      98   
2      Based on 86 Critic Reviews                      98   
3      Based on 73 Critic Reviews                      97   
4      Based on 87 Critic Reviews                      97   
...                           ...                     ...   
13384  Based on 14 Critic Reviews                      19   
13385   Based on 7 Critic Reviews                      18   
13386  Based on 19 Critic Reviews                      17   
13387  Based on 15 Critic Reviews                      17   
13388   Based on 7 Critic Reviews                      17   

                  user_review_count user_review_score  \
0      Based on 10,273 User Ratings               9.1   
1         Based on 574 User Ratings               7.8   
2       Based on 4,462 User Ratings               8.3   
3       Based on 4,027 User Ratings               9.1   
4       Based on 2,857 User Ratings               9.0   
...                             ...               ...   
13384     Based on 468 User Ratings               1.4   
13385      Based on 10 User Ratings               6.1   
13386      Based on 65 User Ratings               1.6   
13387     Based on 122 User Ratings               3.2   
13388      Based on 60 User Ratings               5.0   

                                         developer  \
0                    Developer:           Nintendo   
1                       Developer:           Namco   
2              Developer:           Rockstar North   
3                    Developer:           Nintendo   
4          Developer:           Nintendo EAD Tokyo   
...                                            ...   
13384               Developer:           Eutechnyx   
13385            Developer:           Pronto Games   
13386                 Developer:           Gravity   
13387                 Developer:           Poisoft   
13388  Developer:           Backbone Entertainment   

                               publisher  \
0                    Publisher: Nintendo   
1                       Publisher: Namco   
2              Publisher: Rockstar Games   
3                    Publisher: Nintendo   
4                    Publisher: Nintendo   
...                                  ...   
13384             Publisher: Deep Silver   
13385               Publisher: D2C Games   
13386                     Publisher: CFK   
13387                 Publisher: Poisoft   
13388  Publisher: Backbone Entertainment   

                                            platforms                 genres  \
0                                     ['Nintendo 64']  ['Open-World Action']   
1      ['Dreamcast', 'iOS (iPhone/iPad)', 'Xbox 360']        ['3D Fighting']   
2                 ['PlayStation 3', 'Xbox 360', 'PC']  ['Open-World Action']   
3                                             ['Wii']      ['3D Platformer']   
4                                             ['Wii']      ['3D Platformer']   
...                                               ...                    ...   
13384             ['PlayStation 3', 'PC', 'Xbox 360']   ['Action Adventure']   
13385                     

baris yang memiliki nilai nan saya hapus untuk meningkatkan performa model


---
fungsi dropna mengambil baris mana yang memiliki nilai nan, kemudian menghapusnya


In [9]:
df_clean.isnull().sum()

Unnamed: 0                 0
name                       0
metacritic_review_count    0
metacritic_review_score    0
user_review_count          0
user_review_score          0
developer                  0
publisher                  0
platforms                  0
genres                     0
esrb                       0
must_play                  0
dtype: int64

Terlihat tidak adanya lagi missing value disetiap kolom

## Mengecek duplikasi

In [10]:
df_clean.duplicated().sum()

np.int64(0)

tidak terdapat baris dengan nilai yang sama

---

fungsi duplicated mengambil baris yang memiliki duplikasi, dan sum menghitung jumlah duplikasi yang ada

## Menghapus kolom

In [11]:
df_clean = df_clean.drop('Unnamed: 0', axis=1)

saya menghapus kolom Unnamed:0 yang tidak sengaja terbuat ketika memanggil file csv

---
fungsi drop akan langsung menghapus nama kolom dalam argumen, dalam hal ini 'Unnamed: 0'


## Mengecek Nama Unik

In [12]:
df_clean['genres'].unique()

array(["['Open-World Action']", "['3D Fighting']", "['3D Platformer']",
       "['Skating']", "['FPS']", "['Football Sim']",
       "['Linear Action Adventure']", "['Survival']", "['Western RPG']",
       "['Compilation']", "['Visual Novel']", "['Action RPG']",
       "['Auto Racing Sim']", "['Action Adventure']", "['2D Platformer']",
       "['JRPG']", "['Third Person Shooter']", "['Auto Racing']",
       "['4X Strategy']", "['2D Fighting']", "['Real-Time Strategy']",
       "['Arcade']", "['Command RTS']", "['Skiing']", "['Sandbox']",
       "['Soccer Sim']", "['Rhythm']", '["2D Beat-\'Em-Up"]',
       "['Basketball Sim']", "['Roguelike']", "['MMORPG']",
       "['Hockey Sim']", "['Virtual Life']", "['Turn-Based Tactics']",
       "['Tactical FPS']", "['Tennis']", "['Metroidvania']",
       "['Aircraft Sim']", "['Aircraft Combat Sim']",
       "['Vehicle Combat Sim']", "['Baseball Sim']",
       "['Point-and-Click']", "['Golf']", "['Action Puzzle']",
       "['Basketball']", "['Danci

Menjabarkan nilai unique dari genres, sebagai pertimbangan masukan untuk fungsi rekomendasi nanti

In [13]:
df_clean['publisher'].unique()

array(['Publisher: Nintendo', 'Publisher: Namco',
       'Publisher: Rockstar Games', ..., 'Publisher: CFK',
       'Publisher: Poisoft', 'Publisher: Backbone Entertainment'],
      dtype=object)

Mengecek nilai unique dari publisher, untuk pertimbangan masukan fungsi rekomendasi nanti

---
fungsi unique mengambil nilai unik dari setiap kolom yang disebutkan


## Mengurutkan berdasarkan Genre Games

In [14]:
df_clean.sort_values(by='genres', inplace=True)

Mengurutkan berdasarkan genre, untuk mempermudah menemukan game dengan genre yang sama

---
fungsi sort.values akan mengurutkan values berdasarkan kolom yang dipilih, dalam hal ini genres


## Menormalisasi Data

In [15]:
# Menormalisasi kolom metacritic_review_count
df_clean['metacritic_review_count'] = df_clean['metacritic_review_count'].str.replace('Based on', '')
df_clean['metacritic_review_count'] = df_clean['metacritic_review_count'].str.replace('Critic Reviews', '')

# Menormalisasi kolom user_review_count
df_clean['user_review_count'] = df_clean['user_review_count'].str.replace('Based on', '')
df_clean['user_review_count'] = df_clean['user_review_count'].str.replace('User Ratings', '')

#Menormalisasi kolom Developer
df_clean['developer'] = df_clean['developer'].str.replace('Developer: ', '')

# Menormalisasi kolom publiser
df_clean['publisher'] = df_clean['publisher'].str.replace('Publisher: ', '')

# Menormalisasi Kolom esrb
df_clean['esrb'] = df_clean['esrb'].str.replace('Rated', '')

Membersihkan kolom dari string, berfungsi agar kolom hanya memiliki nilai numeric, untuk mempersiapkan analysis. Jikalau kolom tersebut akan dipakai nantinya

---
fungsi str.replace menganti value strin menjadi value yang lain. dalam hal ini menganti kata dengan nilai kosong


## Merubah Type Data

In [16]:
# Merubah type data metacritic_review_count
df_clean['metacritic_review_count']=df_clean['metacritic_review_count'].astype(int)

# merubah type data metacritic_review_score
df_clean['metacritic_review_score']=df_clean['metacritic_review_score'].astype(int)

# merubah type data user_review_count
df_clean['user_review_count'] = df_clean['user_review_count'].str.replace(',', '').astype(float)

# merubah type data user_review_score
df_clean['user_review_score']=df_clean['user_review_score'].astype(float)

Mengubah type data kolom menjadi int dan float, agar lebih sesuai dengan isi kolom sekarang

---
fungsi astype menganti type kolom menjadi type yang kita inginkan


In [17]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9902 entries, 7265 to 8704
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   name                     9902 non-null   object 
 1   metacritic_review_count  9902 non-null   int64  
 2   metacritic_review_score  9902 non-null   int64  
 3   user_review_count        9902 non-null   float64
 4   user_review_score        9902 non-null   float64
 5   developer                9902 non-null   object 
 6   publisher                9902 non-null   object 
 7   platforms                9902 non-null   object 
 8   genres                   9902 non-null   object 
 9   esrb                     9902 non-null   object 
 10  must_play                9902 non-null   int64  
dtypes: float64(2), int64(3), object(6)
memory usage: 928.3+ KB


Menampilkan informasi dataset yang sudah dibersihkan, terlihat kalau dataset hanya memiliki 9902 baris yang tersisa.
metacritic_review_count dan metacritic_review_score sudah berubah menjadi int.
user_review_count dan user_review_score sudah berubah menjadi float

## Membuat kolom Combinasi


In [18]:
df_clean['combined_features']= df_clean['name'].fillna('') + '' + df_clean['developer'].fillna('') + '' + df_clean['publisher'].fillna('') + '' + df_clean['genres'].fillna('') + '' + df_clean['esrb'].fillna('')

kolom kombinasi berguna untuk menghitung nilai cosine_similarity nantinya.

---
kolom combined_features akan berisi value kombinasi dari kolom yang dipilih, setiap value akan dipisahkan dengan '' melalui fungsi fillna


## menormalisasi kolom kombinasi

In [19]:
def clean_text(text):
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)  # hapus tanda baca
    text = re.sub(r'\d+', '', text)      # hapus angka
    return text

df_clean['combined_features']=df_clean['combined_features'].apply(clean_text)

Melakukan normalisasi terhadap kolom combined_features, dengan mengubah menjadi lower case, membersihkan tanda baca, dan angka untuk mempersiapkan untuk digunakan dalam tf idf nantinya.

---

fungsi text.lower mengubah setiap text menjadi lower case.
fungsi re.sub menghapus tanda baca dan angka untuk setiap text

# Model Development dengan Content Based Filtering

## TF-IDF Vectorizer

In [20]:
tfidf = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf.fit_transform(df_clean['combined_features'])

Mempersiapkan agar kolom bisa digunakan dalam sistem rekomendasi menggunakan tf-idf


---

**TfidfVectorizer(stop_words='english'**) memanggil object tf-idf dari library sklearn. **TF-IDF** berguna untuk mengetahui seberapa penting suatu kata dalam setiap dokumen, **stop_words=english** membuat model mengabaikan stop words umum dalam bahasa inggriss
fungsi **fit** belajar mengenail vocabulary dari kolom combined_features, untuk mengetahui nilai unique dan distribusi didalam kolom
fungsi **transform**, untuk mengconvert data kedalam tf-idf matriks.

## Menghitung cosine similarity

In [21]:
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)
print('Ukuran matriks cosine similarity:', cosine_sim.shape)

Ukuran matriks cosine similarity: (9902, 9902)


cosine_sim, berfungsi untuk menghitung similarity antara 2 arrays, dalam hal ini, saya menghitung cosine similarity dalam setiap permainan.
fungsi **shape** mencetak ukuran dari matriks dalam hal ini kesamaan antara setiap game dengan game yang lain, karena tfidf_matrix dipanggil 2 kali.
nilai 9902 mewakili baris dan kolom, juga dapat berarti jumlah game dalam dataset yang sudah dibersihkan

## Membuat fungsi rekomendasi berdasarkan genre

In [22]:
def recommend_games_by_genre(genre, top_n=5):
    # Filter the dataframe by the specified genre
    genre_games = df_clean[df_clean['genres'].str.lower().str.contains(genre.lower(), na=False)]

    if genre_games.empty:
        print(f"No games found for genre: {genre}")
        return pd.DataFrame(columns=['genres', 'name'])

    # Sort the games by metacritic_review_score in descending order
    recommended_games = genre_games.sort_values(by='metacritic_review_score', ascending=False).head(top_n)

    return recommended_games[['genres', 'name']]

fungsi rekomendasi menggunakan 2 parameter, variabel yang akan digunakan untuk menampung inputan, dan top_n=5 akan menampilkan 5 data teratas dari inputan.
fungsi **str.lower** mengubah value dari kolom genres menjadi huruf kecil.
fungsi **str.lower** mengecek apakah string yang dimasukkan tersedia di kolom genres. yang juga sudah diubah ke huruf kecil.
fungsi na=False akan mengangap valued nan sebagai tidak ada
**if genre_games.empty** mengecek apakah data yang dimasukkan tidak ada didalam kolom
**recommended_games** mengurutkan variabel genre_games berdasrkan kolom metacritic_review_score secara Menurun
**fungsi return** akan mengembalikan nilai value, dalam hal ini genres dan name

In [23]:
recommend_games_by_genre("JRPG", top_n=5)

genres                  name
46   ['JRPG']       Persona 5 Royal
60   ['JRPG']          Chrono Cross
50   ['JRPG']      Final Fantasy IX
63   ['JRPG']  Metaphor: ReFantazio
104  ['JRPG']             Persona 5

fungsi recommend_games_by_genre akan mengirimkan 2 parameter, yakni input string, dan berapa jumlah baris yang ditampilkan

## Membuat fungsi rekomendasi berdasarkan publisher

In [24]:
def recommend_games_by_publisher(publisher, top_n=5):
    # Filter the dataframe by the specified publisher
    publisher_games = df_clean[df_clean['publisher'].str.lower().str.contains(publisher.lower(), na=False)]

    if publisher_games.empty:
        print(f"No games found for publisher: {publisher}")
        return pd.DataFrame(columns=['publisher', 'name'])

    # Sort the games by metacritic_review_score in descending order
    recommended_games = publisher_games.sort_values(by='metacritic_review_score', ascending=False).head(top_n)

    return recommended_games[['publisher', 'name']]

fungsi **str.lower** mengubah value dari kolom publisher menjadi huruf kecil.
fungsi **str.lower** mengecek apakah string yang dimasukkan tersedia di kolom publisher. yang juga sudah diubah ke huruf kecil.
fungsi na=False akan mengangap valued NAN sebagai tidak ada
**if publisher_games.empty** mengecek apakah data yang dimasukkan tidak ada didalam kolom publisher
**recommended_games** mengurutkan variabel publisher_games berdasarkan kolom metacritic_review_score secara Menurun
**fungsi return** akan mengembalikan nilai value, dalam hal ini publisher dan name

In [25]:
recommend_games_by_publisher("Adult Swim", top_n=5)

publisher                 name
1747  Adult Swim            Duck Game
2751  Adult Swim  Battle Chef Brigade
4528  Adult Swim           Headlander
7026  Adult Swim       Death's Gambit
7136  Adult Swim           Pool Panic

fungsi diatas bekerja dengan cara yang sama seperti genre.
contohnya ketika saya mengetikkan Adult Swim, maka fungsi akan mencari 5 games teratas yang memiliki publisher yang sama, yakni Adult Swim

# Evaluasi

## Pilih Genre


In [31]:
selected_genre = 'FPS'

Memilih genre yang ingin kita check kemiripannya

In [33]:
game_index = df_clean[df_clean['genres'].str.contains(selected_genre, case=False, na=False)].index[0]

fungsi **str.contains()** mencari konten yang mengandung string tertentu. ini diperlukan karena kolom genre berbentuk list. karena itu tidak bisa jika hanya memasukkan input saya, sistem harus mencari content string didalam list

## Dapatkan Skor Kemiripan

In [34]:
similarity_scores = list(enumerate(cosine_sim[game_index]))

mengambil skor kemiripan dari matriks, dan merubahnya menjadi pasangan indeks dan skor
cosine_sim = merupakan matriks similarity yang telah dihitung.

[game_index] : mengakses baris dalam cosine_sim berdasarkan indeks.

enumarate():mengembalikan objek enumerasi, yang akan menghasilkan pasangan indeks dan nilai

## Urutkan Skor Kemiripan

In [35]:
sorted_similar_games = sorted(similarity_scores, key=lambda x: x[1], reverse=True)

berfungsi untuk mengurutkan games berdasarkan scores similarity yang didapatkan


---
sorted() mengembalikan list baru yang berisi semua item

key=lambda x: x[1]: berfungsi sebagai skor indeks untuk melakukan pengurutan

reverse=True: berfungsi mengembalikan nilai secara Descending, dari besar ke kecil


## Tampilkan Genre yang Mirip

In [40]:
print(f"Top 10 similar genre for '{selected_genre}':")
for i, (index, score) in enumerate(sorted_similar_games[1:11]):
    game_name = df_clean.iloc[index]['name']
    print(f"{i+1}. {game_name} (Similarity Score: {score:.4f})")

Top 10 similar genre for 'FPS':
1. Diablo II: Lord of Destruction (Similarity Score: 0.8478)
2. Diablo (Similarity Score: 0.6771)
3. Diablo III (Similarity Score: 0.6327)
4. Diablo IV (Similarity Score: 0.6187)
5. Diablo III: Ultimate Evil Edition (Similarity Score: 0.5338)
6. Diablo III: Rise of the Necromancer (Similarity Score: 0.5081)
7. Diablo III: Reaper of Souls (Similarity Score: 0.5078)
8. Diablo II: Resurrected (Similarity Score: 0.5027)
9. Diablo III: Eternal Collection (Similarity Score: 0.3438)
10. Starcraft II: Legacy of the Void (Similarity Score: 0.2100)


fungsi print diatas mengembalikan isi berdasarkan inputan yang diberikan

Perintah for akan melakukan pengulangan berdasarkan daftar game yang sudah diurutkan

df_clean.iloc[index]['name'] berfungsi untuk mengambil nama game dari dataset

print(f"{i+1}. {game_name} (Similarity Score: {score:.4f})") berfungsi menampilkan nama games apa saja yang memiliki kemiripan genre